src/main/java/com/limechain/network/protocol/blockannounce/BlockAnnounceService.java

```diff
@@ -19,4 +24,20 @@ public void sendHandshake(Host us, PeerId peer) {
             log.warning("Error sending handshake request to peer " + peer);
         }
     }
+
+    /**
+     * Sends a Block Announce message to a peer. If there is no initiator stream opened with the peer,
+     * sends a handshake instead.
+     *
+     * @param us     our host object
+     * @param peerId message receiver
+     */
+    public void sendBlockAnnounceMessage(Host us, PeerId peerId, byte[] encodedBlockAnnounceMessage) {
+        Optional.ofNullable(connectionManager.getPeerInfo(peerId))
+                .map(p -> p.getBlockAnnounceStreams().getInitiator())
+                .ifPresentOrElse(
+                        stream -> new BlockAnnounceController(stream).sendBlockAnnounceMessage(encodedBlockAnnounceMessage),
+                        () -> sendHandshake(us, peerId)
+                );
```

Comment:
Don't we want the `isPresent` check to happen before `.map`?

Reply:
There is no need to check isPresent() before .map(), because .map() is only invoked if the Optional is present.